### DISCUSSION TITLE
Frequency Modulation with Analog Output

### DISCUSSION THREAD

**Discussion by cvanbibber** [pschatzmann/arduino-audio-tools#494] [Q&A] [answered]
Hi,
I am using the streams-url-mp3-analog and maximilian FM1 examples to attempt to create a system that will stream an mp3 file from a url and modulate it with a constant 40khz sine wave carrier frequency to be used as a signal for an ultrasonic transducer array.

However I am having trouble as the FM maximilian function seems to only support I2SStream as an output, whereas I am looking to use an AnalogAudioStream output on gpio 25/26. Is there any way to merge these two examples or set up a function which outputs an FM analog signal after an mp3 file has been decoded?

Thanks very much, and any help would be appreciated!

**Comment by pschatzmann** [comment, marked as answer]
I have never tried, but I would expect that you should be able to use any type of output stream.
Why do you think AnalogAudioStream is not working ?

However I would have my doubts that you can reach 40khz. 
What sampling rate are you using ?

**Reply by cvanbibber** [reply]
Thanks for your reply!

I managed to get a fairly poor sine wave with about 40khz using a sample rate of 4410000. I dont think I fully understand how sample rate works, but the default 44.1k produced a very rough sine wave. I then have a sinewave.setFrequency(10000) in the void setup after sinewave.begin(1, 4410000, N_B4). I could set up either a 40khz sine or square wave but I think a sine wave would be ideal.

I have attached my code since I realized I hadn't previously.
Right now the main error seems to be the encodedaudiostream dec is ambiguous, as there is both the maximilian and analogaudiostream using the "out" output. I am unsure how to fix this but any help would be appreciated.

Also my void play function is using both the sine wave generated by SineWaveGenerator and a MaxiOsc oscillator. It is giving an error but I am not sure if I can use this or not. I might be able to create another MaxiOsc oscillator and have that set at a constant 40khz, while the other changes based on the decoded mp3 file, however I am also not sure how to add this.



```
 // One way of thinking about FM synthesis is to see it as vibrato.
// You make a pitch, then vary it up and down at some rate.
// You can change the speed of the pitch variation (modulation frequency), and also the amount of variation (modulation index).
// In FM, usually only one of the waveforms - the carrier that provides the initial pitch - is sent to the output.
// The frequency of the the carrier wave is continually adjusted at a rate equal to the frequency of the second wave (the modulator).
// So at any given point in time, the frequency of the carrier can increase by an amount equal to the current amp of the modulator.
// This has some interesting effects.

#include "AudioTools.h"
#include "AudioLibs/MaximilianDSP.h"
#include "AudioCodecs/CodecMP3Helix.h"

// Define Arduino output
URLStream url("ssid","pass"); // ssid, password
typedef int16_t sound_t;                                   // sound will be represented as int16_t (with 2 bytes)
uint16_t sample_rate=4410000;
uint8_t channels = 1; 
SineWaveGenerator<sound_t> sineWave(32000);                // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<sound_t> sound(sineWave);             // Stream generated from sine wave
AnalogAudioStream out;
EncodedAudioStream dec(&out, new MP3DecoderHelix()); // Decoding stream
StreamCopy copier(dec, url); // copy url to decoder
Maximilian maximilian(out);
// Maximilian
maxiOsc myOtherSine;//One Maximilian oscillator


void setup() {//some inits
  // setup logging
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup Audio output
  auto cfg = out.defaultConfig(TX_MODE);
  cfg.sample_rate = sample_rate;
  out.begin(cfg);
  sineWave.begin(channels, sample_rate, N_B4);
  sleep(1);
  sineWave.setFrequency(10000); // this is about 1/4 of the actual frequency
  Serial.println("started...");
  maximilian.begin(cfg);

  dec.setNotifyAudioChange(out);
  dec.begin();

  url.begin("http://stream.srg-ssr.ch/m/rsj/mp3_128","audio/mp3");
}

void play(float *output) {
    
    // In this example, the 'myOtherSine.sinewave' is at an amplitude of 1, it's original amplitude.
    // This is pretty simple and not too useful.
    //output[0]=mySine.sinewave(440*myOtherSine.sinewave(1));
    
    // Perhaps you should comment out the above line and uncomment the below one instead
    // It shows how the frequency of the carrier is altered by ADDING a second waveform to its frequency value.
    // The carrier frequency is 440, and the modulation frequency is 1.
    // It also shows how the modulation index works. In this case the modulation index is 100
    // Try adjusting the modolation index. Also, try altering the modulation frequency.
    output[0]=sineWave(10000+(myOtherSine.sinewave(1)*100));
    output[1]=output[0];

}

// In complex FM systems you can have lots of modulators stacked together in interesting ways, and theoretically this can make any sound.
// John Chowning is the guy you probably want to talk to about that.

// Arduino loop
void loop() {
  copier.copy();
    maximilian.copy();
}
 ```

**Comment by pschatzmann** [comment]
In fact, the maximum bandwidth of a sampled waveform is determined exactly by its sampling rate; the maximum frequency representable in a sampled waveform is termed its Nyquist frequency, and is equal to one half the sampling rate. 
So if you want to generate a 40khz sine wave you need a sampling rate of at least 80khz...

I think you would need to write the decoded data into some buffer (e.g. a CallbackBufferedStream) and use it modulate the output.

**Comment by TylerCheetah80** [comment]
Hello, I am a partner on the project.

I was wondering what buffer you would recommend using and how I can write decoder values to said buffer.

Also, I have been having trouble finding the documentation for CallbackBufferedStream, would I just call it on the buffer itself though?

Any insight is appreciated, thanks!

**Comment by pschatzmann** [comment]
- Link to documentatin: https://pschatzmann.github.io/arduino-audio-tools/html/classaudio__tools_1_1_callback_buffered_stream.html
- Example: https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-audiokit-multioutput-server/streams-audiokit-multioutput-server.ino

It is a Stream API over a queue: You would use this as output for the EncodedAudioStream and in the play method you would update the values every n samples where n is sampling rate osciallator / sampling rate audio to modulate the signal:
```
CallbackBufferedStream<uint8_t> queue(buffer_size, buffer_count); 
int16_t samples[2]; // left & right samples
...
queue.readBytes((uint8_t*) samples, sizeof(int16_t)*2); 
output[0]=sineWave(10000+(samples[0]/scale)));
output[1]=sineWave(10000+(samples[0]/scale)));

 ```

Don't forget to scale the values correctly. In your example you are using a float (in the range of -1 to 1), so you would divide the samples e.g with 327.

**Reply by cvanbibber** [reply]
Thanks for this, this was very helpful!

We are having a problem with our esp32 (esp32-s2-solo native usb) in uploading this code, is there a physical limitation to this chip resulting in us not being able to perform these functions using it or would we be able to bypass it and if so how?

Thanks!

**Reply by pschatzmann** [reply]
If you don't give me any error message, it will be pretty difficult to help. MP3 uses a lot of progmem, so did you set the Partition Scheme to Huge App ?
In addition as far as I can remember the S2 does not support analog output via I2S, so AnalogAudioStream might not work...
